File: src/lb.zig
const std = @import("std");
const builtin = @import("builtin");
const posix = std.posix;

const MAX_BACKENDS: usize = 8;
const DEFAULT_BACKLOG: u32 = 4096;
const DEFAULT_ACCEPT_BATCH: u32 = 128;

const Iovec = extern struct {
    base: ?*anyopaque,
    len: usize,
};

const Msghdr = extern struct {
    name: ?*anyopaque,
    namelen: u32,
    iov: ?*Iovec,
    iovlen: usize,
    control: ?*anyopaque,
    controllen: usize,
    flags: c_int,
};

const Cmsghdr = extern struct {
    len: usize,
    level: c_int,
    type: c_int,
};

const SCM_RIGHTS: c_int = 1;
const SOL_SOCKET: c_int = 1;
const MSG_NOSIGNAL: c_int = 0x4000;
const MSG_DONTWAIT: c_int = 0x40;
const SOCK_SEQPACKET: u32 = 5;
const TCP_NODELAY: c_int = 1;
const TCP_DEFER_ACCEPT: c_int = 9;
const TCP_QUICKACK: c_int = 12;

const MAX_PREFIX: usize = 4096;

inline fn cmsgAlign(len: usize) usize {
    const a: usize = @sizeOf(usize);
    return (len + a - 1) & ~(a - 1);
}
inline fn cmsgSpace(len: usize) usize {
    return cmsgAlign(len) + cmsgAlign(@sizeOf(Cmsghdr));
}
inline fn cmsgLen(len: usize) usize {
    return cmsgAlign(@sizeOf(Cmsghdr)) + len;
}

extern fn sendmsg(sockfd: c_int, msg: *const Msghdr, flags: c_int) isize;
extern fn chmod(path: [*:0]const u8, mode: c_uint) c_int;

pub fn main() !void {
    const port_cstr = std.c.getenv("LB_PORT");
    const port: u16 = if (port_cstr) |p| (std.fmt.parseInt(u16, std.mem.span(p), 10) catch 9999) else 9999;
    const backlog_cstr = std.c.getenv("LB_BACKLOG");
    const backlog: u32 = if (backlog_cstr) |p| (std.fmt.parseInt(u32, std.mem.span(p), 10) catch DEFAULT_BACKLOG) else DEFAULT_BACKLOG;
    const batch_cstr = std.c.getenv("LB_ACCEPT_BATCH");
    const accept_batch: u32 = if (batch_cstr) |p| (std.fmt.parseInt(u32, std.mem.span(p), 10) catch DEFAULT_ACCEPT_BATCH) else DEFAULT_ACCEPT_BATCH;

    const sockets_cstr = std.c.getenv("API_SOCKETS") orelse {
        std.debug.print("[lb] set API_SOCKETS=/sock/api1.sock,/sock/api2.sock\n", .{});
        return error.MissingEnv;
    };
    const sockets = std.mem.span(sockets_cstr);

    var backend_fds: [MAX_BACKENDS]c_int = undefined;
    var nb: usize = 0;
    {
        var it = std.mem.splitScalar(u8, sockets, ',');
        while (it.next()) |raw_p| {
            const p = std.mem.trim(u8, raw_p, " ");
            if (p.len == 0) continue;
            if (nb >= MAX_BACKENDS) break;
            try waitForPath(p);
            var conn_tries: u32 = 0;
            backend_fds[nb] = while (conn_tries < 600) : (conn_tries += 1) {
                if (connectUds(p)) |fd| break fd else |_| {}
                std.time.sleep(100 * std.time.ns_per_ms);
            } else return error.BackendConnectTimeout;
            std.debug.print("[lb] connected to {s} fd={d}\n", .{ p, backend_fds[nb] });
            nb += 1;
        }
    }
    if (nb == 0) return error.NoBackends;

    const lfd = try posix.socket(posix.AF.INET, posix.SOCK.STREAM | posix.SOCK.NONBLOCK | posix.SOCK.CLOEXEC, 0);
    defer posix.close(lfd);
    try posix.setsockopt(lfd, posix.SOL.SOCKET, posix.SO.REUSEADDR, &std.mem.toBytes(@as(c_int, 1)));
    try posix.setsockopt(lfd, posix.SOL.SOCKET, posix.SO.REUSEPORT, &std.mem.toBytes(@as(c_int, 1)));
    _ = std.c.setsockopt(lfd, std.c.IPPROTO.TCP, @intCast(TCP_DEFER_ACCEPT), &std.mem.toBytes(@as(c_int, 1)), 4);

    const addr: posix.sockaddr.in = .{
        .family = posix.AF.INET,
        .port = std.mem.nativeToBig(u16, port),
        .addr = 0,
        .zero = .{0} ** 8,
    };
    try posix.bind(lfd, @ptrCast(&addr), @sizeOf(posix.sockaddr.in));
    try posix.listen(lfd, @intCast(@min(backlog, std.math.maxInt(u31))));
    std.debug.print("[lb] listening :{d} batch={d} backends={d}\n", .{ port, accept_batch, nb });

    var rr: usize = 0;
    while (true) {
        var accepted: u32 = 0;
        var got_one: bool = false;
        while (accepted < accept_batch) : (accepted += 1) {
            const cfd = posix.accept(lfd, null, null, posix.SOCK.NONBLOCK | posix.SOCK.CLOEXEC) catch break;
            got_one = true;
            _ = posix.setsockopt(cfd, std.c.IPPROTO.TCP, @intCast(TCP_NODELAY), &std.mem.toBytes(@as(c_int, 1))) catch {};
            _ = std.c.setsockopt(cfd, std.c.IPPROTO.TCP, @intCast(TCP_QUICKACK), &std.mem.toBytes(@as(c_int, 1)), 4);

            var prefix_buf: [MAX_PREFIX]u8 = undefined;
            const prefix_len = readReadyPrefix(cfd, &prefix_buf);

            const start = rr;
            var attempt: usize = 0;
            while (attempt < nb) : (attempt += 1) {
                const idx = (start + attempt) % nb;
                if (sendFdWithBytes(backend_fds[idx], cfd, prefix_buf[0..prefix_len])) {
                    rr = (idx + 1) % nb;
                    break;
                } else |_| {}
            }
            posix.close(cfd);
        }
        if (!got_one) {
            var pfd = [_]posix.pollfd{.{ .fd = lfd, .events = posix.POLL.IN, .revents = 0 }};
            _ = posix.poll(&pfd, 60_000) catch {};
        }
    }
}

fn waitForPath(path: []const u8) !void {
    var tries: u32 = 0;
    while (tries < 600) : (tries += 1) {
        var z_buf: [256]u8 = undefined;
        if (path.len >= z_buf.len) return error.PathTooLong;
        @memcpy(z_buf[0..path.len], path);
        z_buf[path.len] = 0;
        if (std.fs.cwd().accessZ(@ptrCast(&z_buf), .{})) {
            return;
        } else |_| {}
        std.time.sleep(100 * std.time.ns_per_ms);
    }
    return error.Timeout;
}

fn connectUds(path: []const u8) !c_int {
    const fd_pos: posix.fd_t = try posix.socket(posix.AF.UNIX, SOCK_SEQPACKET | posix.SOCK.CLOEXEC, 0);
    errdefer posix.close(fd_pos);
    const sndbuf: c_int = 256 * 1024;
    try posix.setsockopt(fd_pos, posix.SOL.SOCKET, posix.SO.SNDBUF, &std.mem.toBytes(sndbuf));
    var addr: posix.sockaddr.un = .{ .family = posix.AF.UNIX, .path = undefined };
    @memset(&addr.path, 0);
    const n = @min(addr.path.len - 1, path.len);
    @memcpy(addr.path[0..n], path[0..n]);
    const addrlen: posix.socklen_t = @intCast(@offsetOf(posix.sockaddr.un, "path") + n + 1);
    try posix.connect(fd_pos, @ptrCast(&addr), addrlen);
    return @intCast(fd_pos);
}

fn readReadyPrefix(fd: posix.fd_t, buf: []u8) usize {
    // Drain whatever is currently in the kernel buffer. We consume the bytes
    // (no MSG_PEEK), so we MUST forward them to the worker — anything we leave
    // behind would still arrive via epoll on the duplicated fd in the worker.
    var len: usize = 0;
    while (len < buf.len) {
        const got = posix.recv(fd, buf[len..], MSG_DONTWAIT) catch |err| switch (err) {
            error.WouldBlock => return len,
            else => return len,
        };
        if (got == 0) return len;
        len += got;
    }
    return len;
}

fn sendFdWithBytes(uds_fd: c_int, client_fd: posix.fd_t, prefix: []const u8) !void {
    var len_hdr: u16 = @intCast(prefix.len);
    var dummy: u8 = 1;
    // First iov: 2-byte length prefix (always present, even when 0).
    // Second iov: prefix bytes (may be empty, then we pad with a dummy byte for SEQPACKET reception parity).
    var iov_buf: [2]Iovec = .{
        .{ .base = @ptrCast(&len_hdr), .len = @sizeOf(u16) },
        .{ .base = if (prefix.len > 0) @constCast(@ptrCast(prefix.ptr)) else @ptrCast(&dummy), .len = if (prefix.len > 0) prefix.len else 1 },
    };
    const iovlen: usize = if (prefix.len > 0) 2 else 1;

    const CTRL_LEN = cmsgSpace(@sizeOf(c_int));
    var ctrl_buf: [CTRL_LEN]u8 align(@alignOf(Cmsghdr)) = undefined;

    var msg = Msghdr{
        .name = null,
        .namelen = 0,
        .iov = &iov_buf[0],
        .iovlen = iovlen,
        .control = @ptrCast(&ctrl_buf),
        .controllen = CTRL_LEN,
        .flags = 0,
    };

    const cmsg: *Cmsghdr = @ptrCast(@alignCast(&ctrl_buf));
    cmsg.len = cmsgLen(@sizeOf(c_int));
    cmsg.level = SOL_SOCKET;
    cmsg.type = SCM_RIGHTS;
    const data_off = cmsgAlign(@sizeOf(Cmsghdr));
    const data_ptr: *c_int = @ptrCast(@alignCast(&ctrl_buf[data_off]));
    data_ptr.* = @intCast(client_fd);
    msg.controllen = cmsg.len;

    while (true) {
        const r = sendmsg(uds_fd, &msg, MSG_NOSIGNAL);
        if (r > 0) return;
        const e = std.posix.errno(r);
        if (e == .INTR) continue;
        return error.SendFailed;
    }
}

fn sendFd(uds_fd: c_int, client_fd: posix.fd_t) !void {
    var dummy: u8 = 1;
    var iov = Iovec{ .base = @ptrCast(&dummy), .len = 1 };
    const CTRL_LEN = cmsgSpace(@sizeOf(c_int));
    var ctrl_buf: [CTRL_LEN]u8 align(@alignOf(Cmsghdr)) = undefined;

    var msg = Msghdr{
        .name = null,
        .namelen = 0,
        .iov = &iov,
        .iovlen = 1,
        .control = @ptrCast(&ctrl_buf),
        .controllen = CTRL_LEN,
        .flags = 0,
    };

    const cmsg: *Cmsghdr = @ptrCast(@alignCast(&ctrl_buf));
    cmsg.len = cmsgLen(@sizeOf(c_int));
    cmsg.level = SOL_SOCKET;
    cmsg.type = SCM_RIGHTS;
    const data_off = cmsgAlign(@sizeOf(Cmsghdr));
    const data_ptr: *c_int = @ptrCast(@alignCast(&ctrl_buf[data_off]));
    data_ptr.* = @intCast(client_fd);
    msg.controllen = cmsg.len;

    while (true) {
        const r = sendmsg(uds_fd, &msg, MSG_NOSIGNAL | MSG_DONTWAIT);
        if (r > 0) return;
        const e = std.posix.errno(r);
        if (e == .INTR) continue;
        return error.SendFailed;
    }
}

File: src/lb.c
#define _GNU_SOURCE
#include <arpa/inet.h>
#include <errno.h>
#include <fcntl.h>
#include <netinet/in.h>
#include <netinet/tcp.h>
#include <poll.h>
#include <signal.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/socket.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <sys/un.h>
#include <time.h>
#include <unistd.h>

#define MAX_BACKENDS 32
#define DEFAULT_ACCEPT_BATCH 128
#define DEFAULT_BACKLOG 4096
#define MAX_PREFIX 4096

/*
 * Layout of the SCM_RIGHTS message sent to each worker:
 *   iov[0] = uint16_t prefix_len  (little-endian on x86_64; native order, matching
 *            what the Zig worker uses)
 *   iov[1] = prefix_len bytes already drained from the client socket (or a single
 *            dummy byte when prefix_len == 0, so SEQPACKET keeps a non-zero
 *            payload — recvmsg returns r >= sizeof(u16)).
 *
 * This matches src/main.zig::onCtrlRecv which expects len_hdr in iov[0] and
 * the prefix buffer in iov[1].
 */
typedef struct {
    int fd;
    uint16_t len_hdr;
    uint8_t prefix[MAX_PREFIX];
    uint8_t dummy;
    struct iovec iov[2];
    union {
        struct cmsghdr cm;
        char buf[CMSG_SPACE(sizeof(int))];
    } control;
    struct msghdr msg;
    struct cmsghdr *cmsg;
} backend_t;

static int getenv_int(const char *name, int fallback) {
    const char *v = getenv(name);
    if (!v || !*v) return fallback;
    int parsed = atoi(v);
    return parsed > 0 ? parsed : fallback;
}

static int connect_backend(const char *path) {
    int fd = socket(AF_UNIX, SOCK_SEQPACKET | SOCK_CLOEXEC, 0);
    if (fd < 0) return -1;
    int sndbuf = 256 * 1024;
    setsockopt(fd, SOL_SOCKET, SO_SNDBUF, &sndbuf, sizeof(sndbuf));
    struct sockaddr_un addr = {0};
    addr.sun_family = AF_UNIX;
    strncpy(addr.sun_path, path, sizeof(addr.sun_path) - 1);
    if (connect(fd, (struct sockaddr *)&addr, sizeof(addr)) < 0) {
        close(fd);
        return -1;
    }
    return fd;
}

static void init_backend(backend_t *b, int fd) {
    memset(b, 0, sizeof(*b));
    b->fd = fd;
    b->dummy = 1;
    /* iov[0] always points at len_hdr; iov[1] base/len rewritten per send. */
    b->iov[0].iov_base = &b->len_hdr;
    b->iov[0].iov_len = sizeof(uint16_t);
    b->iov[1].iov_base = &b->dummy;
    b->iov[1].iov_len = 1;

    b->msg.msg_iov = b->iov;
    b->msg.msg_iovlen = 2;
    b->msg.msg_control = b->control.buf;
    b->msg.msg_controllen = sizeof(b->control.buf);
    b->cmsg = CMSG_FIRSTHDR(&b->msg);
    b->cmsg->cmsg_level = SOL_SOCKET;
    b->cmsg->cmsg_type = SCM_RIGHTS;
    b->cmsg->cmsg_len = CMSG_LEN(sizeof(int));
}

static int wait_for_socket(const char *path) {
    int tries = 0;
    while (tries++ < 600) {
        struct stat st;
        if (stat(path, &st) == 0) return 0;
        struct timespec ts = { .tv_sec = 0, .tv_nsec = 100 * 1000 * 1000 };
        nanosleep(&ts, NULL);
    }
    return -1;
}

/*
 * Inline /ready short-circuit. Peeks the first bytes of the request without
 * consuming them and, if it looks like a GET /ready (no path suffix), replies
 * 200 OK directly from the LB and closes. Saves the SCM_RIGHTS cross + worker
 * wakeup for every Docker/contest health check. Reference: whereisanzi
 * crates/lb/src/main.rs::peek_is_health.
 */
static const char READY_RESP[] =
    "HTTP/1.1 200 OK\r\nContent-Length: 2\r\nConnection: close\r\n\r\nok";

static int try_serve_ready(int cfd) {
    char buf[64];
    ssize_t n = recv(cfd, buf, sizeof(buf), MSG_PEEK | MSG_DONTWAIT);
    if (n < (ssize_t)sizeof("GET /ready") - 1) return 0;
    if (memcmp(buf, "GET /ready", 10) != 0) return 0;
    /* Confirm boundary: next char must be ' ', '?', or end-of-buffer-not-yet-arrived
     * (we already peeked 10 bytes). The lb never serves other GET /ready* routes,
     * so accept space, '?', '\r' as the boundary. */
    char c = buf[10];
    if (c != ' ' && c != '?' && c != '\r' && c != '\n') return 0;
    const char *resp = READY_RESP;
    size_t left = sizeof(READY_RESP) - 1;
    while (left > 0) {
        ssize_t s = send(cfd, resp, left, MSG_NOSIGNAL);
        if (s > 0) { resp += s; left -= (size_t)s; continue; }
        if (s < 0 && errno == EINTR) continue;
        break;
    }
    return 1;
}

/*
 * Drain whatever bytes are already sitting in the TCP receive buffer for cfd.
 * Non-blocking; returns the number of bytes consumed (0..MAX_PREFIX). Anything
 * we read here MUST be forwarded — the worker will not see these bytes by
 * re-reading the fd, because we just consumed them.
 */
static size_t drain_ready_prefix(int cfd, uint8_t *buf) {
    /* Single recv: TCP_DEFER_ACCEPT guarantees data is present, and the worker's
     * onRecv loops until EAGAIN to pick up any tail. Saves a syscall per request,
     * which is hot on the LB (constrained to ~0.02 CPU under throttling). */
    ssize_t r = recv(cfd, buf, MAX_PREFIX, MSG_DONTWAIT);
    return r > 0 ? (size_t)r : 0;
}

static int send_fd_with_flags(backend_t *dst, int cfd, size_t prefix_len, int flags) {
    /* Rebuild controllen each send: MSG_CTRUNC etc. can stomp it. */
    dst->msg.msg_controllen = sizeof(dst->control.buf);
    dst->cmsg->cmsg_level = SOL_SOCKET;
    dst->cmsg->cmsg_type = SCM_RIGHTS;
    dst->cmsg->cmsg_len = CMSG_LEN(sizeof(int));
    memcpy(CMSG_DATA(dst->cmsg), &cfd, sizeof(int));

    dst->len_hdr = (uint16_t)prefix_len;
    if (prefix_len > 0) {
        dst->iov[1].iov_base = dst->prefix;
        dst->iov[1].iov_len = prefix_len;
    } else {
        dst->iov[1].iov_base = &dst->dummy;
        dst->iov[1].iov_len = 1;
    }
    dst->msg.msg_iovlen = 2;

    for (;;) {
        ssize_t r = sendmsg(dst->fd, &dst->msg, MSG_NOSIGNAL | flags);
        if (r > 0) return 0;
        if (r < 0 && errno == EINTR) continue;
        return -1;
    }
}

static int send_fd(backend_t *dst, int cfd, size_t prefix_len) {
    return send_fd_with_flags(dst, cfd, prefix_len, MSG_DONTWAIT);
}

static int send_fd_blocking(backend_t *dst, int cfd, size_t prefix_len) {
    return send_fd_with_flags(dst, cfd, prefix_len, 0);
}

static int parse_backends(const char *env, char *paths[MAX_BACKENDS]) {
    int n = 0;
    char *tmp = strdup(env);
    char *save = NULL;
    char *tok = strtok_r(tmp, ",", &save);
    while (tok && n < MAX_BACKENDS) {
        /* trim surrounding whitespace */
        while (*tok == ' ' || *tok == '\t') tok++;
        size_t L = strlen(tok);
        while (L > 0 && (tok[L-1] == ' ' || tok[L-1] == '\t')) tok[--L] = 0;
        if (L > 0) paths[n++] = strdup(tok);
        tok = strtok_r(NULL, ",", &save);
    }
    free(tmp);
    return n;
}

int main(int argc, char **argv) {
    (void)argc;
    (void)argv;
    signal(SIGPIPE, SIG_IGN);

    int port = 9999;
    if (getenv("LB_PORT")) port = atoi(getenv("LB_PORT"));
    int backlog = getenv_int("LB_BACKLOG", DEFAULT_BACKLOG);
    int accept_batch = getenv_int("LB_ACCEPT_BATCH", DEFAULT_ACCEPT_BATCH);
    const char *socks_env = getenv("API_SOCKETS");
    if (!socks_env || !*socks_env) socks_env = "/sock/api1.sock,/sock/api2.sock";

    char *paths[MAX_BACKENDS] = {0};
    int nb = parse_backends(socks_env, paths);
    if (nb <= 0) {
        fprintf(stderr, "[lb] no backends\n");
        return 2;
    }

    static backend_t backends[MAX_BACKENDS];
    for (int i = 0; i < nb; i++) {
        fprintf(stderr, "[lb] waiting %s\n", paths[i]);
        if (wait_for_socket(paths[i]) < 0) {
            fprintf(stderr, "[lb] timeout waiting %s\n", paths[i]);
            return 3;
        }
        int fd = -1;
        for (int t = 0; t < 100; t++) {
            fd = connect_backend(paths[i]);
            if (fd >= 0) break;
            struct timespec ts = { .tv_sec = 0, .tv_nsec = 100 * 1000 * 1000 };
            nanosleep(&ts, NULL);
        }
        if (fd < 0) {
            fprintf(stderr, "[lb] connect failed %s\n", paths[i]);
            return 4;
        }
        init_backend(&backends[i], fd);
        fprintf(stderr, "[lb] connected %s (fd=%d)\n", paths[i], fd);
    }

    int lfd = socket(AF_INET, SOCK_STREAM | SOCK_NONBLOCK | SOCK_CLOEXEC, 0);
    if (lfd < 0) { perror("socket"); return 5; }
    int on = 1;
    setsockopt(lfd, SOL_SOCKET, SO_REUSEADDR, &on, sizeof(on));
    setsockopt(lfd, SOL_SOCKET, SO_REUSEPORT, &on, sizeof(on));
    setsockopt(lfd, IPPROTO_TCP, TCP_DEFER_ACCEPT, &on, sizeof(on));

    struct sockaddr_in addr = {0};
    addr.sin_family = AF_INET;
    addr.sin_addr.s_addr = htonl(INADDR_ANY);
    addr.sin_port = htons(port);
    if (bind(lfd, (struct sockaddr *)&addr, sizeof(addr)) < 0) { perror("bind"); return 6; }
    if (listen(lfd, backlog) < 0) { perror("listen"); return 7; }

    fprintf(stderr, "[lb] listening :%d backlog=%d batch=%d backends=%d\n",
            port, backlog, accept_batch, nb);

    int rr = 0;
    for (;;) {
        int accepted = 0;
        while (accepted < accept_batch) {
            int cfd = accept4(lfd, NULL, NULL, SOCK_NONBLOCK | SOCK_CLOEXEC);
            if (cfd < 0) {
                if (errno == EINTR) continue;
                break; /* EAGAIN/EWOULDBLOCK or other → exit batch */
            }
            accepted++;
            int one = 1;
            setsockopt(cfd, IPPROTO_TCP, TCP_NODELAY, &one, sizeof(one));
            setsockopt(cfd, IPPROTO_TCP, TCP_QUICKACK, &one, sizeof(one));

            /* Short-circuit /ready at the LB so Docker/contest health checks
             * never wake an API worker. TCP_DEFER_ACCEPT guarantees data is
             * already in the receive buffer before accept returns. */
            if (try_serve_ready(cfd)) { close(cfd); continue; }

            int target = rr;
            rr = (rr + 1) % nb;

            /* Drain bytes already present (TCP_DEFER_ACCEPT means the request
             * is usually fully buffered) directly into the backend's prefix
             * staging area, then send fd + prefix in a single SCM_RIGHTS msg. */
            size_t prefix_len = drain_ready_prefix(cfd, backends[target].prefix);

            if (send_fd(&backends[target], cfd, prefix_len) != 0) {
                (void)send_fd_blocking(&backends[target], cfd, prefix_len);
            }
            close(cfd);
        }
        if (accepted == 0) {
            struct pollfd pfd = { .fd = lfd, .events = POLLIN, .revents = 0 };
            poll(&pfd, 1, -1);
        }
    }
}
